src/consumer/__tests__/consumeMessages.spec.js

```diff
@@ -609,6 +609,57 @@ describe('Consumer', () => {
     })
   })
 
+  it('discards messages received when pausing while fetch is in-flight', async () => {
+    consumer = createConsumer({
+      cluster: createCluster(),
+      groupId,
+      maxWaitTimeInMs: 1000,
+      logger: newLogger(),
+    })
+
+    const messages = Array(10)
+      .fill()
+      .map(() => {
+        const value = secureRandom()
+        return { key: `key-${value}`, value: `value-${value}` }
+      })
+    await producer.connect()
+    await producer.send({ acks: 1, topic: topicName, messages })
+
+    await consumer.connect()
+
+    await consumer.subscribe({ topic: topicName, fromBeginning: true })
+
+    const sleep = value => waitFor(delay => delay >= value)
+    let offsetsConsumed = []
+
+    const eachBatch = async ({ batch, heartbeat }) => {
+      for (const message of batch.messages) {
+        offsetsConsumed.push(message.offset)
+      }
+
+      await heartbeat()
+    }
+
+    consumer.run({
+      eachBatch,
+    })
+
+    await waitForConsumerToJoinGroup(consumer)
+
+    await waitFor(() => offsetsConsumed.length === messages.length, { delay: 50 })
+    await sleep(50)
+
+    // Hope that we're now in an active fetch state? Something like FETCH_START might help
+    const seekedOffset = offsetsConsumed[Math.floor(messages.length / 2)]
+    consumer.pause([{ topic: topicName }])
+    await producer.send({ acks: 1, topic: topicName, messages }) // trigger completion of fetch
+
+    await sleep(200)
```

Comment:
Rather than sleeping an arbitrary amount and hoping we've consumed the message by then, use the `waitFor` utility.

Reply:
The point is that no message has been consumed, that nothing has happened, so that makes a `waitFor` a bit trickier! Happy to hear if you've got any suggestions for that :)

Reply:
Ah, I should have read it more properly. Perhaps there's an event you could wait for having happened, such as `END_BATCH_PROCESS` to make sure that a fetch has happened.

Reply:
Ended up going with `FETCH`, which seems to capture the use case correctly. Looked into `END_BATCH_PROCESS`, which worked for the unfixed implementation, but never triggered in the first as the filtered response was an empty batch, which get skipped!